### PR TITLE
Chris/AIP 6471 - Add single script to test multiple use cases of flow-triggering feature

### PR DESCRIPTION
**Pull request by cavandervoort**
This PR adds a single python script that iterates through various test cases of the flow-triggering feature.

Currently, there are two scripts used for testing, but the parameters must be changed manually each time to test a particular aspect of the flow-triggering feature. I plan to keep these two scripts, as they run more quickly and are useful in debugging while developing. The script added in this PR will iterate through multiple use cases and will test both proper and improper (error-causing) use of the flow-triggering feature.

**UPDATE:** I deleted the single test script b/c unittest allows us to run a single test at a time.

**UPDATE:** While the tests all work and the feature passes every test, there is annoying output from the test notifying the user about ResourceWarnings. I added a TODO in the test_trigger_live_run.py file so that we later investigate this and fix it. I also created [Jira Task AIP-6586](https://zbrt.atl.zillow.net/browse/AIP-6586) to address this issue.

### CODE REVIEW

File: metaflow/plugins/argo/argo_live_run.py
Comment (cloudw) [inline metaflow/plugins/argo/argo_live_run.py:65]: Any reason why 12 hour of time out is needed? I am not sure what default timeout should be used. I'm most trying to understand your thought process.

File: metaflow/plugins/argo/argo_live_run.py
Comment (cloudw) [inline metaflow/plugins/argo/argo_live_run.py]: (minor and not part of this PR) `for key, value in parameters.items()` slightly simplify the code.

File: metaflow/plugins/argo/argo_live_run.py
Comment (cloudw) [inline metaflow/plugins/argo/argo_live_run.py]: Can you explain what this does in comment?

File: metaflow/plugins/argo/argo_live_run.py
Comment (cloudw) [inline metaflow/plugins/argo/argo_live_run.py]: Can you also add doc string for why `loop_counter` is needed? I think your design is fine for checking status more frequently than status report, but you want to make your design explicit to developers who do not share the same context.

File: metaflow/plugins/argo/test/test_trigger_live_run.py
Comment (cloudw) [inline metaflow/plugins/argo/test/test_trigger_live_run.py]: Could you please restructure this file according to sample unit test usage here? https://docs.python.org/3/library/unittest.html#basic-example

Eventually we'd like to have some unit test package (for example `pytest`) running your test scripts.

File: metaflow/plugins/argo/test/test_single_run.py
Comment (cloudw) [inline metaflow/plugins/argo/test/test_single_run.py]: Please clean up commented out lines here before merging.

Comment (cloudw) [reply metaflow/plugins/argo/test/test_single_run.py]: Actually, should this file be removed entirely since you have all tests in `test_trigger_live_run.py`?

File: metaflow/plugins/argo/test/test_trigger_live_run.py
Comment (cloudw) [inline metaflow/plugins/argo/test/test_trigger_live_run.py:31]: Does all integration tests here rely on the corresponding argo workflow template to exist in the cluster first? 

Ideally the test should:
- Not rely on cluster state for testing. Which requires each test run to 
  1. setup test resources (upload argo workflow template)
  2. run tests
  3. clean up temporary test resources (delete workflows template) every time the test is ran
- Allow multiple test runs at the same time. This requires that workflow template version doesn't conflict (blocked by workflow template versioning strategy, or an alternative way that fulfill the goal for temporary test flows - you can identify each workflow template using some unique id like test trigger time /  test pipeline id etc)


Feel free to create backlog tickets in Jira for them if you feel like these will take significant time / you can't get to them by end of the internship.

**Review by cloudw — CHANGES_REQUESTED** [review]
Thanks for creating the test script!

The test script on its own looks good. Still I suggest you align with python conventions of writing test cases in https://docs.python.org/3/library/unittest.html, so that existing test packages can be leveraged for a nice test output. For [example](https://gitlab.zgtools.net/analytics/artificial-intelligence/ai-platform/sdks/zillow-metaflow/-/jobs/14420559), pytest will provide counts over number of test passed / failed, print stderr for failed tests, and report test coverages. 

Some comments are about code not in this PR - sorry for not catching them earlier.

File: metaflow/plugins/argo/argo_live_run.py
Comment (cloudw) [reply metaflow/plugins/argo/argo_live_run.py]: Will this logic result in sleeping for negative amount of time?

File: metaflow/plugins/argo/argo_live_run.py
Comment (cavandervoort) [reply metaflow/plugins/argo/argo_live_run.py:65]: It was just a guess as to what's best. I've heard of some flows taking multiple hours to run, so I wanted to be conservative and not cut off any flows before they are finished. I kept in a TODO about this because I don't know what the best default setting is.

Comment (cavandervoort) [reply metaflow/plugins/argo/argo_live_run.py:65]: I added a note about this in our 1:1 backlog

File: metaflow/plugins/argo/argo_live_run.py
Comment (cavandervoort) [reply metaflow/plugins/argo/argo_live_run.py]: It can't go negative because x % 5 will never be greater than 5.

I added this because it takes time to lookup status, which made the status appear slightly less frequently than one per minute. I added this comment:

> the mod operator is used to keep each loop at exactly 5 seconds,
> so status can be printed every minute w/ help from loop_counter

Is this sufficient? 

I could also just make it 5s each time to avoid any confusion. It's not a big deal. On larger flows, you'd occasionally see something like:

> Time waited: 78 minutes out of a possible 720
> Time waited: 79 minutes out of a possible 720
> Time waited: 80 minutes out of a possible 720
> Time waited: 82 minutes out of a possible 720
> Time waited: 83 minutes out of a possible 720

Skipping from 80 to 82 is a little odd, but not a big deal.

What do you think I should do? My sense is that we keep the code the same, with the explanatory comment.

File: metaflow/plugins/argo/argo_live_run.py
Comment (cavandervoort) [reply metaflow/plugins/argo/argo_live_run.py]: I added comment:

> tracks loop count so status printed every minute

Is this ok?

File: metaflow/plugins/argo/test/test_single_run.py
Comment (cavandervoort) [reply metaflow/plugins/argo/test/test_single_run.py]: Yes, I think removing is best. Will do.

File: metaflow/plugins/argo/argo_live_run.py
Comment (cloudw) [inline metaflow/plugins/argo/argo_live_run.py]: minor: missing a "minute" in the end. Maybe `minutes out of {wait_timeout} minutes limit.`

File: metaflow/plugins/argo/argo_live_run.py
Comment (cloudw) [reply metaflow/plugins/argo/argo_live_run.py]: I'd try to make the design / intention more explicit by either adding more doc str or renaming a few variable. 

Maybe instead of say sleep(5), you can define local var called `status_check_interval: int = 5`, and a local var `status_report_interval = 60` to be compared with `time.time() - start_time`.

File: metaflow/plugins/argo/argo_live_run.py
Comment (cloudw) [reply metaflow/plugins/argo/argo_live_run.py]: Oh I see what you are doing now. That's some nice math work!

I think for this feature it's not important to be off for a few seconds. Constant 5 seconds wait is fine.

**Review by cloudw — CHANGES_REQUESTED** [review]

File: metaflow/plugins/argo/argo_live_run.py
Comment (cloudw) [reply metaflow/plugins/argo/argo_live_run.py:65]: I think 12 hour is good for not having user worrying about setting a different value. The downside is, if the triggered flow stuck in runner state for any reason and customer didn't expect run time to be that long, this function will wait for a long time

For cases where user needs a shorter timeout, they have better idea of how long a wait is needed.

File: metaflow/plugins/argo/test/test_trigger_live_run.py
Comment (cavandervoort) [reply metaflow/plugins/argo/test/test_trigger_live_run.py:31]: I made a Jira task for this: [AIP-6585 - Make tests not reliant on cluster state and allow concurrent tests](https://zbrt.atl.zillow.net/browse/AIP-6585)

File: metaflow/plugins/argo/argo_live_run.py
Comment (cavandervoort) [reply metaflow/plugins/argo/argo_live_run.py:65]: Sounds good. I'll leave it at 12 hours for the default value

File: metaflow/plugins/argo/test/live_run_failure_flow.py
Comment (cloudw) [inline metaflow/plugins/argo/test/live_run_failure_flow.py:17]: Sleep can be removed since you only test `is_running` against simple flow. The same applies for the other 3 flows.

File: metaflow/plugins/argo/test/live_run_failure_flow.py
Comment (cloudw) [inline metaflow/plugins/argo/test/live_run_failure_flow.py:18]: `middle_step` can be removed. Apart from shorter code it saves a slight amount of time and cost per test run.

File: metaflow/plugins/argo/test/live_run_flow_triggering_flow.py
Comment (cloudw) [inline metaflow/plugins/argo/test/live_run_flow_triggering_flow.py:55]: Can move all of this to start step and remove this step. The same applies for a few other flows.

**Review by cloudw — APPROVED** [review]
Looks great in general. Approved.

Some minor feedback from slightly perfectionism side of me. 🙂

File: metaflow/plugins/argo/test/live_run_failure_flow.py
Comment (cavandervoort) [reply metaflow/plugins/argo/test/live_run_failure_flow.py:17]: I will remove the 5s wait from flows other than simple and failure flow. Failure flow does test is_running while it's running in the ["test_failure_flow" test](https://github.com/zillow/metaflow/blob/69c48fbe1459521219005d55931ccf57ad5021da/metaflow/plugins/argo/test/test_trigger_live_run.py#L180).

**Pull request merged**
